File: 0070-climbing-stairs/0070-climbing-stairs.py
class Solution:
    def climbStairs(self, n: int) -> int:   

        my_dict = {}

        def recursion(n):
            if n == 0:
                return 0
            if n == 1:
                return 1

            if n == 2:
                return 2

            #if its computed
            if n in my_dict:
                return my_dict[n]

            #if its not computed
            my_dict[n] = recursion(n - 1) + recursion(n - 2)

            #return
            return my_dict[n]

        return recursion(n)
        

File: 0070-climbing-stairs/0070-climbing-stairs.cpp
class Solution {
public:
    int climbStairs(int n) 
    {
        vector<int> memo(n + 1, -1);
        return climbStairs2(0, n, memo);    
    }

    int climbStairs2(int steps, int n, vector<int>& memo)
    {
        //Base case
        if (steps > n)
        {
            return 0;
        }

        if (steps == n)
        {
            return 1;
        }

        //Check memo
        if (memo[steps] != -1)
        {
            return memo[steps];
        }

        //Choice 1
        int one = climbStairs2(steps + 1, n, memo);

        //Choice 2
        int two = climbStairs2(steps + 2, n, memo);

        //Add to meom
        memo[steps] = one + two;

        return one + two;
    }
};